### ISSUE TITLE
No way to apply policies to user-owned projects?

### ISSUE THREAD

**Issue by llamasoft** [egnyte/gitlabform#106]
There doesn't appear to be a way to apply settings to user-owned projects (i.e. projects under `username/project-name`) because they exist under a "user" namespace, not a "group" namespace.  I see that there is a `get_all_projects` function that might be a workaround for this, but it doesn't appear to be used by the code.

Being able to apply configurations to user-owned projects would be handy to make sure that users aren't putting sensitive information into publicly accessible projects.

**Comment by gdubicki** [comment]
That's surprising, I have been testing GitLabForm multiple times on my user-owned projects...

Are you using token of an admin user of your GitLab instance?

**Comment by gdubicki** [comment]
Oh wait, you want to apply config to *other users'* projects, not just your own?

**Comment by llamasoft** [comment]
Correct.  We have an "admin" account whose token we use for running automated configuration things.  The admin user themself doesn't down any projects, but we need them to apply configurations to all projects, including other users.

**Comment by gdubicki** [comment]
This turned out to be a bit harder to implement than I thought because GitLab's API is missing an advertised feature of listing users with projects - see https://gitlab.com/gitlab-org/gitlab/-/issues/217960 ...

**Comment by gdubicki** [comment]
When you get a list of users with projects somehow there is another problem - no API for getting all user's project. 🙄 GitLab issue: https://gitlab.com/gitlab-org/gitlab/-/issues/16149

**Comment by UncleStark** [comment]
I tried to apply the configuration to all projects of the user, whose key is specified in the config. In the config there was a section common _settings. Run gitlabform ALL. As a result, the configuration was applied only to projects that belong to any groups. I started it with the -d switch and saw a listing of all user's projects in the log. I parsed them and added "project_settings" to the conafig, then ran ALL _DEFINED. After that, everything worked out as it should. plz tell me why ALL key does not work.

**Comment by gdubicki** [comment]
ALL doesn’t work in this case because it tries to get a list of all groups & projects and you cannot do that because of aforementioned gitlab bug. ALL_DEFINED takes explicitly listed groups & projects, so it does work. It may be a valid workaround for some cases but not for the OP request, where he wants to apply settings to all users’ projects.

**Comment by UncleStark** [comment]
I understand it. my question is why the list of all projects was in the log but not seen by the application

**Comment by gdubicki** [comment]
Oh, sorry, I missed that @UncleStark . I will check it out. What GitLab version are you using? Maybe this makes a difference. I assune that Gitlabform is the latest?

**Comment by UncleStark** [comment]
GitLabForm version: 1.17.0.1 (the latest)
Connected to GitLab version: 13.4.3 (6689552c479)

**Comment by gdubicki** [comment]
You are right, @UncleStark : we can get the list of users' projects from GitLab! Now the question is how to allow Gitlabform to benefit from this?

I would like to keep backwards compatibility, so I would not make the user-owned projects included when applying "common" level settings.

I would rather add a new configuration level, like "personal_project_settings". A Gitlabform user will be able to easily YAML anchors to apply the same config for them as for "common", but its their choice.

Opinions?

**Comment by UncleStark** [comment]
I think this is the perfect solution
both the wolves have eaten much and the sheep have not been touched
